### PR TITLE
Validate involvedObject.Namespace matches event.Namespace

### PR DESCRIPTION
**Pull request by liggitt**
Fixes https://github.com/kubernetes/kubernetes/issues/16036

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30533)

### CODE REVIEW

**Comment by liggitt** [comment]
Re-work of https://github.com/kubernetes/kubernetes/pull/16146

**Comment by liggitt** [comment]
cc @derekwaynecarr @lavalamp @feihujiang from earlier related PR

File: pkg/api/validation/events.go [hunk 1]
```diff
@@ -17,29 +17,64 @@ limitations under the License.
...
-	// to api.NamespaceDefault, so we accept that too, but "" is preferred.
-	if (event.InvolvedObject.Kind == "Node" || event.InvolvedObject.Kind == "PersistentVolume") &&
-		event.Namespace != api.NamespaceDefault &&
-		event.Namespace != "" {
-		allErrs = append(allErrs, field.Invalid(field.NewPath("involvedObject", "namespace"), event.InvolvedObject.Namespace, "not allowed for node"))
+
+	// Make sure event.Namespace and the involvedObject.Namespace agree
+	if len(event.InvolvedObject.Namespace) == 0 {
+		// event.Namespace must also be empty (or "default", for compatibility with old clients)
+		if event.Namespace != "" && event.Namespace != api.NamespaceDefault {
```
Comment (smarterclayton) [inline pkg/api/validation/events.go]: Use `api.NamespaceNone` for the empty string.

Comment (liggitt) [inline pkg/api/validation/events.go]: done

**Comment by liggitt** [comment]
@k8s-bot unit test this issue #30462

**Pull request merged**